core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4

```diff
@@ -292,6 +296,59 @@ sampleType
     | SYSTEM
     ;
 
+patternRecognitionRelation
+    : aliasedRelation (patternRecognition (AS? identifier columnAliases?)?)?
+    ;
+
+patternRecognition
+    : MATCH_RECOGNIZE '('
+        (PARTITION BY partition+=expression (',' partition+=expression)*)?
+        (ORDER BY sortItem (',' sortItem)*)?
+        (MEASURES measureDefinition (',' measureDefinition)*)?
+        rowsPerMatch?
+        rowPatternCommon
+      ')'
+    ;
+
+measureDefinition
+    : expression AS identifier
+    ;
+
+rowsPerMatch
+    : ONE ROW PER MATCH
+    | ALL ROWS PER MATCH emptyMatchHandling?
+    ;
+
+emptyMatchHandling
+    : SHOW EMPTY MATCHES
+    | OMIT EMPTY MATCHES
+    | WITH UNMATCHED ROWS
+    ;
+
+rowPatternCommon
```

Comment:
Why is this a separate rule? It should be part of patternRecognition to avoid an artificial AST node that serves little purpose

Reply:
This is the part of the grammar that MATCH_RECOGNIZE and pattern matching in WINDOW share. I separated it for easier reuse in the future. Also, the spec defines it this way.

Reply:
I would not over-optimize for the future unless it's very clear how that will need to look like. We can always refactor later if necessary.

Also, we don't need to match the exact structure of the syntactic rules in the SQL spec -- sometimes they are not the best way of organizing things, or they structure they that way to convey some semantic aspects that would otherwise be hard to express.

Reply:
I'll see if inlining `rowPatternCommon` is helpful or not.
Adding pattern matching in WINDOW is not "some future", but I truly hope to add it soon. It mostly requires supporting grammar and analysis. Execution is done.

Reply:
`rowPatternCommon` is now inlined.